Programmer en Scala 3: Prérequis

Iltotore opened the pull request:
This is a small introduction to Scala 3.

Most of Scala articles target experienced people and assumes their audience already knows about programming basis. This article aims to provide a beginer-friendly introduction to Scala using an educational tool named `scala-cli`.

florentm35 commented on langages/scala/programmer-en-scala-3-prerequis.md:
```diff
@@ -0,0 +1,64 @@
+
+Scala, dont le nom veut dire "Scalable language" (Langage qui s'adapte à la taille du projet), est un langage orienté objet et fonctionnel créé par Martin Odersky. Scala est un langage JVM, c'est à dire tournant sur la Java Virtual Machine, le logiciel utilisé pour exécuter les programmes écrits avec le langage Java. De ce fait, beaucoup d'entreprises utilisant Java utilisent Scala pour booster l ...
+
+Scala est un langage dit "general purpose" (qui n'est pas centré sur un domaine précis) qui est le plus souvent utilisé pour développer pour créer des systèmes concurrents (qui peut gérer plusieurs tâches "en même temps") pouvant traiter de grandes quantité de données/beaucoup d'utilisateurs en même temps. Le langage est utilisé par des entreprises telles que Netflix ou Twitter.
```
Il va falloir que tu choisisse : "pour développer pour créer"

Equinox4 replied:
ah, bien vu, je l'ai loupé lors de ma review celui-ci 😄

Iltotore replied:
> Il va falloir que tu choisisse : "pour développer pour créer" 

Effectivement petite erreur de ma part. Je corrige ça une fois de nouveau sur mon PC

Iltotore replied:
Corrigé dans le dernier commit

Elanis reviewed (approved):
Je relis en tant que non connaisseur de scala ou de la JVM, mais LGTM.

Julien00859 commented:
The message on the pull-request is nice, it is a shame it is not on the first commit. When we'll merge this PR all will stay in the git log is the commit, not the PR. The first commit should be amended so that it includes that messages, you can use an interactive rebase for that:


![rebase-step-by-step](https://user-images.githubusercontent.com/8208953/177160311-af1c0e04-31aa-442a-9563-e4c2dca18642.png)

![interactive-rebase](https://user-images.githubusercontent.com/8208953/177160318-a7123cd4-bab6-4393-8834-b04b2cd6d853.png)

Use the opportunity to fixup the two last commits into the `add:` one. We aint gonna merge two `fix:` commits on the very PR that introduce the article.

Note that you'll need to force-push your branch, cool kids use `--force-with-lease`

Julien00859 commented:
> Most of Scala articles target experienced people and assumes their audience already knows about programming basis. This article aims to provide a beginer-friendly introduction to Scala.

I don't agree *at all* that this article is *beginner-friendly*. In the first three sentences you talk about (1) the JVM, (2) Java, (3) Object Oriented Programming and (4) Functional Programming. Not any of those terms is *beginner-friendly*, you need some background knowledge on programming to understand what it is.

A beginner doesn't need to know about the JVM/Java or the OOP/FP paradigms to write its first few programs, the way you introduce them feels very odd. Instead he would be interested to know what Scala is: it is a programming langage, a way to instruct the computer into doing work. He would also be interested to know what are the kind of programs developers usually use Scala for, ditch the "general purpose", they all are, give some precise industry-standard example where Scala really shines.

The very first programming example should be minimalist: ditch the comments, they are noise for the beginner. You can hint a few term in a sentence, e.g.

> Here we wrote our first program, it simply print "Hello world" on screen and exit. To do so we created a main function that call `println` with the text we wanted to print. We'll see in the next chapter what it means.

---

From what I understand, this article will be the first chapter of a series on Scala. I am not sure it is a good idea just to merge each chapter in a separated PR without prior discussion on the structure. IMO there should first be an issue addressing the scope, audience, tone and subject of the matter. Maybe an initial structure should be suggested inside that issue so we better understand how would each article fit with the others in the serie.

Iltotore commented:
> A beginner doesn't need to know about the JVM/Java or the OOP/FP paradigms to write its first few programs, the way you introduce them feels very odd.

You're right. Without Scala CLI I would still introduce Scala as a JVM language because it is required to run it (using its main implementation) but since Scala CLI already abstracts this, it is not relevant anymore.

> He would also be interested to know what are the kind of programs developers usually use Scala for, ditch the "general purpose", they all are, give some precise industry-standard example where Scala really shines.

I already talked about the main domain where Scala is used but it should probably be digged further.

> The very first programming example should be minimalist: ditch the comments, they are noise for the beginner. You can hint a few term in a sentence, e.g.

Seems to be a better idea.

> From what I understand, this article will be the first chapter of a series on Scala. I am not sure it is a good idea just to merge each chapter in a separated PR without prior discussion on the structure. IMO there should first be an issue addressing the scope, audience, tone and subject of the matter. Maybe an initial structure should be suggested inside that issue so we better understand how would each article fit with the others in the serie.

I'm not sure when I will write the next article. I don't know if time matters but I can create an issue with a first "plan".

Iltotore commented:
Everything should be ok now:
- Grouped all related commits (add + edit) into a single one with the new message
- Fixed the typo found by florentm
- Rewrote the introduction to make the article more beginer friendly
- Now use a Scala script `.sc` instead of a `.scala` file to avoid to introduce the main method. The entire code is now `println("Hello World")`

florentm35 commented on langages/scala/programmer-en-scala-3-prerequis.md:
````diff
@@ -0,0 +1,55 @@
+Ouvrez maintenant votre terminal à l'endroit où vous avez créé le fichier et entrez la commande suivante:
+
+```sh
+scala-cli hello.sc
````
Si le cours est débutant friendly, une cause de problème qu'il risque d'avoir souvent c'est les problèmes de path, je ne sait pas si un petit lin ou une petite note sur le sujet ne pourrais pas être rajouté ?

TheElectronWill commented:
Happy to see a post about Scala3 :tada:

Two small remarks:
- It seems unnecessary to distinguish between "UNIX" and non-UNIX systems. What matters to a beginner is that it works everywhere
- Scala CLI is great, but for a true beginner, maybe it would be simpler to work in the browser with [Scastie](https://scastie.scala-lang.org/)? It would be even easier since no installation would be required. Of course, for bigger projects, a local installation is required, but I believe that it won't be necessary before 1 or 2 posts. (Furthermore, no IDE has been suggested to the beginner yet - it would have more fun in scastie than in gedit/notepad :smile:)

Iltotore commented:
> Happy to see a post about Scala3 :tada:
> 
> Two small remarks:
> - It seems unnecessary to distinguish between "UNIX" and non-UNIX systems. What matters to a beginner is that it works everywhere

I agree I could just mention Linux, Mac OS and Windows but I don't think that the current sentence is confusing for beginners. Might be still worth to fix anyway.

> - Scala CLI is great, but for a true beginner, maybe it would be simpler to work in the browser with [Scastie](https://scastie.scala-lang.org/)? It would be even easier since no installation would be required. Of course, for bigger projects, a local installation is required, but I believe that it won't be necessary before 1 or 2 posts. (Furthermore, no IDE has been suggested to the beginner yet - it would have more fun in scastie than in gedit/notepad :smile:)

I agree that Scastie is a great Scala sandbox but :
- You have to be online to use it
- There is no multiple files support. This is not impactful for very first bases but when I will tackle imports, this will be problematic and they'll have to switch to another tool.
- Scala cli is not a build tool. It is more simple and similar to the `python`/`py` command but for Scala.
- When you want to save your work from Scastie, a full SBT project is downloaded which is not relevant at all for a beginner.

Also I didn't suggested any IDE because any editor, even gedit/notepad, is sufficient to learn bases.

TheElectronWill commented:
Oops, sorry for the late reply
It's true that Scastie is limited. Might be worth to mention it at some point? It can be very handy :)
Any editor is sufficient, yes, but without proper syntax highlighting it hurts my eyes :smile: ([notepad++ doesn't support scala out of the box](https://npp-user-manual.org/docs/programing-languages/) :cry: )

Iltotore commented:
> It's true that Scastie is limited. Might be worth to mention it at some point? It can be very handy :)

I think I can do that. Would you mind submit a PR for this change? I barely have the time to do it properly right now.
